Add karateclub implementation of embedding.train.node2vec

paul-tqh-nguyen opened the pull request:
This PR adds an implementation of `node2vec` using the `karateclub` library.

paul-tqh-nguyen committed:
Initial commit of environment.yml, setup.*, versioneer.py

This commit adds the initial implementations of the following files:

	environment.yml
	setup.cfg
	setup.py
	versioneer.py

paul-tqh-nguyen committed:
Add karateclub implementation of embedding.train.node2vec

This commit adds an implementation of node2vec using the karateclub library.

jim22k commented on metagraph_karateclub/plugins/karateclub/algorithms.py:
```diff
@@ -0,0 +1,47 @@
+            q=q,
+            dimensions=embedding_size,
+        )
+        old2canonical = {
```
Does karate club require nodes to be numbered 0..n? The incoming graph should already have valid NodeIDs for the node labels, although they may not be perfectly sequential.

If this is a requirement, you should make a copy of the graph. Modifying the node ids changes the meaning of the graph, so modifying the input graph directly is wrong.

paul-tqh-nguyen replied:
Yes, `karateclub` requires nodes to be numbered sequentially. The code that does that is [here](https://github.com/benedekrozemberczki/karateclub/blob/master/karateclub/estimator.py#L53).

I think `nx.relabel_nodes` makes a copy by default since it has the default parameter `copy=True`. https://github.com/metagraph-dev/metagraph-karateclub/pull/1/commits/149a9a65ddc9bbb84bc910714b7a87f85edce25f makes the copy explicit by explicitly passing in `copy=True`.

jim22k commented on metagraph_karateclub/plugins/karateclub/algorithms.py:
```diff
@@ -0,0 +1,47 @@
+        np_embedding_matrix = trainer.get_embedding()
+        matrix = NumpyMatrix(np_embedding_matrix)
+        node_ids = np.array(list(graph.value.nodes))
+        node2index = NumpyNodeMap(
```
Something is wrong here. You create `node_ids` above, but never use it.

paul-tqh-nguyen replied:
https://github.com/metagraph-dev/metagraph-karateclub/pull/1/commits/2311dc6fc69b209208a30bf62146e7f7db8d1372 removes this variable. 

That was a mistake. I used `node_ids` in my early drafts of the code but clearly did not remember to remove it.

paul-tqh-nguyen committed:
Make graph copying in karateclub_node2vec_train explicit.

When reading the code for karateclub_node2vec_train and running across nx.relabel_nodes, it might not be clear that nx.relabel_nodes returns a copy since copy=True is a default parameter value. This commit makes the code read more explicitly. This commit is intended to be a no-op for karateclub_node2vec_train.

paul-tqh-nguyen committed:
Remove useless variable in karateclub_node2vec_train

karateclub_node2vec_train used to create the variable node_ids and not use it. This was a leftover artifact from refactoring that was incomplete since it missed this. This commit removes this variable.
